File: snippets/python/src/usecases/eventprocessing/user_feed.py
from datetime import timedelta

import restate

from utils import (
    create_post,
    get_post_status,
    update_user_feed,
    SocialMediaPost,
    Status,
)

# <start_here>
user_feed = restate.VirtualObject("UserFeed")


@user_feed.handler()
async def process_post(ctx: restate.ObjectContext, post: SocialMediaPost):
    user_id = ctx.key()

    post_id = await ctx.run_typed(
        "create post", create_post, user_id=user_id, post=post
    )

    while (
        await ctx.run_typed("get status", get_post_status, post_id=post_id)
        == Status.PENDING
    ):
        await ctx.sleep(timedelta(seconds=5))

    await ctx.run_typed("update feed", update_user_feed, user=user_id, post_id=post_id)


# <end_here>

app = restate.app([user_feed])

if __name__ == "__main__":
    import hypercorn
    import hypercorn.asyncio
    import asyncio

    conf = hypercorn.Config()
    conf.bind = ["0.0.0.0:9080"]
    asyncio.run(hypercorn.asyncio.serve(app, conf))

File: snippets/go/usecases/eventprocessing/userfeed.go
package eventprocessing

import (
	restate "github.com/restatedev/sdk-go"
	"time"
)

type SocialMediaPost struct {
	Content  string `json:"content"`
	Metadata string `json:"metadata"`
}

// <start_here>
type UserFeed struct{}

func (UserFeed) ProcessPost(ctx restate.ObjectContext, post SocialMediaPost) error {
	var userId = restate.Key(ctx)

	postId, err := restate.Run(ctx, func(ctx restate.RunContext) (string, error) {
		return CreatePost(userId, post)
	})
	if err != nil {
		return err
	}

	for {
		status, err := restate.Run(ctx, func(ctx restate.RunContext) (string, error) {
			return GetPostStatus(postId), nil
		})
		if err != nil {
			return err
		}
		if status != PENDING {
			break
		}
		if err = restate.Sleep(ctx, 5*time.Second); err != nil {
			// </mark_2>
			return err
		}
	}

	if _, err := restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
		return UpdateUserFeed(userId, postId)
	}); err != nil {
		return err
	}

	return nil
}

// <end_here>
